File: agc033/a.cpp
#include <queue>
#include <vector>
#include <string>
#include <iostream>

using namespace std;


int main() {
    int H {}, W {};
    cin >> H >> W;
    string A[H];
    for (int h=0; h<H; h++) {
        cin >> A[h];
    }

    // initialize first black positions
    queue<vector<int>*> black_list {};
    int **f = new int*[H];
    for (int h=0; h<H; h++) {
        f[h] = new int[W];
        for (int w=0; w<W; w++) {
            if (A[h][w] == '#') {
                f[h][w] = true;
                black_list.push(new vector<int> {h, w, 0});
            } else {
                f[h][w] = false;
            }
        }
    }

    int ans = 0;
    // expand black positions until all positions fill
    while (!black_list.empty()) {
        vector<int> *pos = black_list.front();
        int h = (*pos)[0];
        int w = (*pos)[1];
        int d = (*pos)[2];
        // up
        if (h-1 >= 0 && f[h-1][w] == false) {
            f[h-1][w] = true;
            black_list.push(new vector<int> {h-1, w, d+1});
        }
        // down
        if (h+1 < H && f[h+1][w] == false) {
            f[h+1][w] = true;
            black_list.push(new vector<int> {h+1, w, d+1});
        }
        // left
        if (w-1 >= 0 && f[h][w-1] == false) {
            f[h][w-1] = true;
            black_list.push(new vector<int> {h, w-1, d+1});
        }
        // right
        if (w+1 < W && f[h][w+1] == false) {
            f[h][w+1] = true;
            black_list.push(new vector<int> {h, w+1, d+1});
        }
        ans = d;
        black_list.pop();
        delete pos;
    }

    cout << ans << endl;
}

File: agc033/a.py
(H, W) = map(int, input().split())
A = []
for i in range(H):
    A.append(input())

F = [[False for w in range(W)] for h in range(H)]
blackList = []
for h in range(H):
    for w in range(W):
        if A[h][w] == '#':
            F[h][w] = True
            blackList.append((h, w))

ans = -1
while blackList:
    #print('blackList: ', blackList)
    ans += 1
    nextBlack = []
    for (h, w) in blackList:
        if h-1 >= 0 and F[h-1][w] == False:
            F[h-1][w] = True
            nextBlack.append((h-1, w))
        if h+1 < H and F[h+1][w] == False:
            F[h+1][w] = True
            nextBlack.append((h+1, w))
        if w-1 >= 0 and F[h][w-1] == False:
            F[h][w-1] = True
            nextBlack.append((h, w-1))
        if w+1 < W and F[h][w+1] == False:
            F[h][w+1] = True
            nextBlack.append((h, w+1))
    #print('nextBlack: ', nextBlack)
    blackList = nextBlack

print(ans)
